session_rpcserver.go

```diff
@@ -925,6 +925,52 @@ func (s *sessionRpcServer) AddAutopilotSession(ctx context.Context,
 	// Check that each requested feature is a valid autopilot feature and
 	// that the necessary rules for the feature have been specified.
 	featureRules := make(map[string]map[string]string, len(req.Features))
+
+	// Determine privacy flags to use for session registration.
+	var privacyFlags session.PrivacyFlags
+	if req.PrivacyFlagsSet {
+		// We apply privacy flags from the session request in order to
+		// to be able to set flags resulting from non-standard feature
+		// configurations.
+		privacyFlags, err = session.Deserialize(req.PrivacyFlags)
+		if err != nil {
+			return nil, fmt.Errorf("error deserializing privacy "+
+				"flags (%v) from request: %w",
+				req.PrivacyFlags, err)
+		}
+	} else {
+		// Otherwise, privacyFlags will contain all the combined (ORed)
+		// privacy flags for all requested features with defaults from
+		// the autopilot. This means that if any of the features
+		// includes a less restrictive privacy flag, this will also
+		// apply to features that in principle can be run with better
+		// privacy. Checks for features' privacy flag and previous
+		// session compatibility are done on the autopilot's side for
+		// upgrade flexibility.
+		for f := range req.Features {
+			// Check that the features is known by the autopilot
+			// server.
+			autopilotFeature, ok := autopilotFeatureMap[f]
+			if !ok {
+				return nil, fmt.Errorf("%s is not a features "+
+					"provided by the Autopilot server", f)
+			}
+
+			// Deserialize and check that we know the privacy flags.
+			featurePrivacyFlags, err := session.Deserialize(
+				autopilotFeature.PrivacyFlags,
```

Comment:
ok so I think if we are gonna do the parsing here (and not in the autopilot ListFeatures itself), then we should at least also parse these in `ListFeatures` so that we can set the `RequiresUpgrade` flag accordingly to also be true if the feature has a privacy bit we dont understand. That way the front end can also more easily indicate to the user that they will not be able to register for this feature instead of letting them only find out when they actually attempt to register for the feature.

Reply:
great point, added!